Add gauge limiter class to be used to throttle async case import tasks based on pillow lag

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
This PR adds generic `Gauge` class that can be used to track any value in HQ. It takes in scopes for which it has to track the values for and provides methods to report the metrics and get the values. 
The PR also adds `GaugeLimiter` which will be used to limit the usage based on the values provided by Gauge class. It is done by observing the config that is provided in `GaugeDefintion` method. 

The PR adds `PIllowLagDefinition` class which will be used to throttle the async tasks which are sending stuff to pillow(like case imports) whenever there is a pillow lag greater than the desired value.


Review by commit 🐡 

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-15594
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
NA
## Safety Assurance
The PR does not start throttling as of now, it just provides the base classes and starts reporting metrics for case pillow. 
The PR has adequate tests for all the functions written to ensure that code works as it is supposed to.
### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Have tested the changes locally, would be adding the PR to staging too.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Has adequate test coverage
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

NA
### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
The migrations should be reverted before reverting the PR.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

dannyroberts commented on corehq/project_limits/rate_limiter.py:
```diff
@@ -235,3 +248,23 @@ def _get_rate_definition_dict(rate_definition):
     }
+
+
+class GaugeLimiter:
```
There's a disconnect for me between the genericness of the names `Gauge` and `GuageLimiter` and the specificness (to measuring pillow offsets) of the implementation.

`RateLimiter` is set up generically versus how it is used to implement a specific measurement like the rates of form submissions (https://github.com/dimagi/commcare-hq/blob/1c970475c96197a30f2f850344927adda229160b/corehq/apps/receiverwrapper/rate_limiter.py#L56), and emulating that organization.

Alternatively, if you think we won't have another use case for tracking an evolving value over time like this for the purposes of making throttling decisions, and you think it's not worth making generic, then perhaps you can rename these classes to make clear that they're specific to tracking pillow lag.

I think I prefer the generic approach, because it separates logically generic behavior (defined in the class definition) from the more complex context-specific functions for pulling data etc (passed in when the class is instantiated).

AmitPhulera replied:
Thanks for bringing it up, Danny. I see your point. I started it out as a generic implementation but over the time it became very much specific to the pillows.
I will try to break up the classes to have a generic one and the one specific for the pillows.

AmitPhulera commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,70 @@
+
+    def get_last_reported_time(self):
+        "returns the last time when the lag was reported"
+        return cache.get(self._get_last_reported_cache_key(), datetime.now())
```
> Should this be using datetime.utcnow() (or whatever the acceptable equivalent is these days)?

Yes. It should use utcnow()


> What does it mean if the default is "right now"? Seems like that is not true? Is it fine for this method to lie about when the actual last reported time was?

I was thinking about the first run when the pillow lag was not configured and have clearly not thought it through. I think it should return `None` and that should be handled down in call stack. Thanks for catching that.

AmitPhulera committed:
add a fn to get all the partitions for a kafka topic

AmitPhulera committed:
add abstract base model for Gauge defintions

AmitPhulera committed:
add gauge class with tests

AmitPhulera committed:
add PillowLagGaugeDefinition model

the model will store config to be used in throttling pillows based on lag. The commit also adds helper to get the info out and makes the model avaialble on django admin

AmitPhulera committed:
add the classes that would help in throttling based on gauge values

AmitPhulera committed:
run make migrations

AmitPhulera committed:
add tests for gauge limiter

AmitPhulera committed:
start reporting the lag metrics for case-sql topic

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/gauge-limiter

AmitPhulera commented:
@dannyroberts @millerdev I think this PR is ready for review now. I have force pushed the new commits in the PR itself. Most of the things in it have changed so you may consider it as a new PR and and review on commit by commit basis.

millerdev commented on corehq/project_limits/gauge.py:
```diff
@@ -86,3 +92,95 @@ def avg(self):
+
+    def allow_usage(self):
+        if not self.gauge_definition:
+            return False
```
Looks like this disallows usage if there is no gauge definition, which does not seem like a reasonable default (assuming absence of gauge definition would be a default state). Am I thinking about that correctly?

AmitPhulera replied:
The code is incorrect. It should have returned `True`. During refactoring I think I just updated the function in the child class and missed the base class. Thanks for pointing it out.

AmitPhulera replied:
[582906f](https://github.com/dimagi/commcare-hq/pull/34800/commits/582906f728aea7fec7b69ea71830bcfa7bf7ac2b)

millerdev commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,186 @@
+        if not self.gauge_definition:
+            # No throttling if config to throttle is not set
+            return True
+        if self._is_ideal_throttle_condition() and self._has_pillow_reported_recently():
+            return False
+        return True
```
nit
```suggestion
        return not (self._is_ideal_throttle_condition() and self._has_pillow_reported_recently())
```

millerdev commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,186 @@
+        max_value = self.gauge_definition.max_value
+        average_value = self.gauge_definition.average_value
+
+        if max_value and average_value:
```
Same as above
```suggestion
        if max_value is not None and average_value is not None:
```

millerdev reviewed (commented):
The test failure looks real.
```
AssertionError: Not all Django models are covered by domain dump.
project_limits.PillowLagGaugeDefinition
```

AmitPhulera committed:
add PillowLagGaugeDefinition model to ignore list of domain dump in test

AmitPhulera commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,79 @@
+        return list(all_observed_values.values())
+
+    def _tracked_event_cache_key(self, tracked_key):
+        return f'{self.__class__}-{tracked_key}-gauge-key'
```
> Is this string intended to contain a repr of the type, or the class name? If the latter it should be self.__class__.__name__ or shorter and with less underscores type(self).__name__

It was the latter. Thanks I will update it. 

> Is this Gauge class expected to an abstract base class? If not, why include the word "gauge" twice in this string?

No it can be used as it is. I will remove 'gauge' word  from the key.

AmitPhulera replied:
[6ce128c](https://github.com/dimagi/commcare-hq/pull/34800/commits/6ce128cdb99be3c20e3a0f1fd01de0ba4d3aa107)

AmitPhulera committed:
use class names instead of repr in cache keys and remove redundant text

AmitPhulera committed:
move base class setup calls to top

AmitPhulera commented on corehq/project_limits/tests/test_gauge.py:
```diff
@@ -0,0 +1,212 @@
+        cls.feature_key = CASE_SQL
+        cls.scopes = topics.get_all_kafka_partitons_for_topic(cls.feature_key)
+        cls.case_gauge = case_pillow_lag_gauge
+        return super().setUpClass()
```
[8046407](https://github.com/dimagi/commcare-hq/pull/34800/commits/8046407cb543386d8a62b79dcaba6c895a1f7f33)

AmitPhulera commented on corehq/project_limits/tests/test_gauge.py:
```diff
@@ -96,3 +105,108 @@ def test_avg(self, mock_get_all_kafka_partitons_for_topic):
+            max_value=10,
+            average_value=8
+        )
+        return super().setUpClass()
```
[8046407](https://github.com/dimagi/commcare-hq/pull/34800/commits/8046407cb543386d8a62b79dcaba6c895a1f7f33)

AmitPhulera commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,186 @@
+        max_observed_value = self.gauge.max()
+        avg_observed_value = self.gauge.avg()
+
+        if not (max_observed_value and avg_observed_value):
```
> Not sure if it could ever matter, but it seems safer to explicitly check for None here. What do you think?

That makes sense. I think we we ever set max value of some metric to 0 then this implementation will fail. 

>But this then brings up another question. Imagine a scenario where max_observed_value is not None and avg_observed_value is None and self.gauge_definition.max_value is not None and self.gauge_definition.average_value is None. What would be the correct thing to do in that case?

I don't think it is possible that `max_observed_value` is not None and `avg_observed_value` is None or vice versa because both `average` and `max` functions get values from `Gauge.get_values()` so either both of them will be None or none of them will be `None`. 
But I see your point of having race condition in fetching these values, I think in the case when one of the observed value is `None` then the following condition
```
if not (max_observed_value and avg_observed_value):
```
we will not throttle and will allow the usage and when the next batch will come then hopefully we will not see the race condition again and the throttling will work as expected. 

Does that answer your question or have I misunderstood your concern here?

AmitPhulera commented on corehq/project_limits/migrations/0003_pillowlaggaugedefinition.py:
```diff
@@ -0,0 +1,29 @@
+        migrations.CreateModel(
+            name='PillowLagGaugeDefinition',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
I think so, the commit in first migration seems to 5 years old. I see that the [change](https://docs.djangoproject.com/en/5.0/releases/3.2/#customizing-type-of-auto-created-primary-keys) happened in Django 3.2 which was realeased on Aprli 6th 2021.

AmitPhulera committed:
add explicit None checks

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
Apply suggestions from code review

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera commented on corehq/project_limits/gauge.py:
```diff
@@ -86,3 +92,95 @@ def avg(self):
+        if not self.allow_usage():
+            time.sleep(self.gauge_definition.throttle_for_seconds)
+
+    def allow_usage(self):
```
I will add it.

AmitPhulera replied:
[91107e5](https://github.com/dimagi/commcare-hq/pull/34800/commits/91107e5defe09a1310defbf8f98f5b5d419374f9)

AmitPhulera committed:
add docstrings for allow_usage method

AmitPhulera committed:
fix logic where we were throttling even when throttle config was not set in the base class

millerdev commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,190 @@
+            computed_value = self.gauge.avg()
+        else:
+            computed_value = self.gauge.max()
+        if computed_value < self.gauge_definition.acceptable_value:
+            return True
+        return False
```
nit
```suggestion
        return computed_value < self.gauge_definition.acceptable_value
```

AmitPhulera committed:
Update corehq/project_limits/gauge.py

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
Merge branch 'master' into ap/gauge-limiter

gherceg commented on corehq/ex-submodules/pillowtop/pillow/interface.py:
```diff
@@ -374,13 +374,22 @@ def __record_change_metric_in_datadog(self, metric, change, processing_time=None
+                },
+                multiprocess_mode=MPM_MAX
+            )
+            if change.partition:
```
What is the reason for checking `change.partition` here? It isn't used in the block. I'd assume maybe `dd_topic`?

AmitPhulera replied:
`dd_topic` is created using topic and partition, if there is not partition it just uses topic. But for the case pillow we need to have both the values. This is just an extended check which may or maynot be useful. I added it to be on safer side.

gherceg replied:
Ok makes sense. I think it could be worth adding a comment since it isn't clear in code why you would need a partition to report usage to the pillow lag gauge. What do you think?

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/34800/commits/75bb94a807a847b2290806d4cd3b24d7106e79bc

gherceg commented on corehq/project_limits/tests/test_gauge.py:
```diff
@@ -0,0 +1,212 @@
+        super().setUpClass()
+        cls.feature_key = CASE_SQL
+        cls.scopes = topics.get_all_kafka_partitons_for_topic(cls.feature_key)
+        cls.case_gauge = case_pillow_lag_gauge
```
Is there a particular reason to save the gauge as a class variable?

AmitPhulera replied:
I don't think so 🤔 I guess I had added it before refactoring where this gauge was used by all the tests.

gherceg commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,188 @@
+        max_observed_value = self.gauge.max()
+        avg_observed_value = self.gauge.avg()
+
+        if max_observed_value is None or avg_observed_value is None:
```
Still trying to understand this code, so maybe not but the comment makes me think this should be:
```suggestion
        if max_observed_value is None and avg_observed_value is None:
```

AmitPhulera replied:
Both of these values are calculated from the same source i.e Gauge.get_values which uses cached values. Ideally if one is None then other one would be None too. But there can be a race condition which can result in one of them being None and whenever that happens we don't want to throttle as we would not have adequate values to do so.

gherceg replied:
Ok cool. A suppose a nit would be on the comment then
> No throttling if we don't have any values

Since that gives the impression _neither_  of the values exists, which an OR will not assert. Maybe "Don't throttle unless both observed values are set"

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/34800/commits/75bb94a807a847b2290806d4cd3b24d7106e79bc

gherceg commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,188 @@
+        max_value = self.gauge_definition.max_value
+        average_value = self.gauge_definition.average_value
+
+        if max_value is not None and average_value is not None:
+            # If the max and average values are both set, consider both before throttling
+
+            if max_observed_value > max_value and avg_observed_value > average_value:
+                return True
+        elif max_value is not None:
+            # If only max value is set, throttle based on max value of gauge.
+            if max_observed_value > max_value:
+                return True
+        elif average_value is not None:
+            # If only averable value is set, throttle based on average value gauge.
+            if avg_observed_value > average_value:
+                return True
```
nit: I would clean this up since once in a certain condition, you can return whether true or false.
```suggestion
        if max_value is not None and average_value is not None:
            return max_observed_value > max_value and avg_observed_value > average_value
        elif max_value is not None:
            return max_observed_value > max_value
        elif average_value is not None:
            return avg_observed_value > average_value
```

AmitPhulera replied:
Yeah. This looks much better. Thanks for the suggestion.

gherceg commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,188 @@
+
+    def wait(self):
+        if not self.allow_usage():
+            time.sleep(self.gauge_definition.throttle_for_seconds)
```
```suggestion
            time.sleep(self.gauge_definition.wait_for_seconds)
```

AmitPhulera replied:
Ahh! Good catch. Missed this one.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/34800/commits/d980bd2fa35b0cfa4d635ae69bf8c2a4c3813392

gherceg commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,188 @@
+            return True
+        return not (self._is_ideal_throttle_condition() and self._has_pillow_reported_recently())
+
+    def _is_ideal_throttle_condition(self):
```
Should this method make use of the gauge definition's aggregator type? It looks like it is defaulting to checking both average and max? Which maybe renders the aggregator unused?

AmitPhulera replied:
As per my conversation with Danny we would like to have a different throttle condition for the pillows where we are saying that it should breach both average and max values, so we have a overridden the `allow_usage` but base class uses `aggregator`

gherceg replied:
Ahh I see. I couldn't find aggregator used elsewhere but I see that it was still referenced as `actionable_metric` in the base class. Left a comment on that.

AmitPhulera replied:
Yeah. I saw it when I was going through the code. Have fixed it in the last commit 😅

AmitPhulera committed:
cleanup if else conditions

Co-authored-by: Graham Herceg <gherceg@dimagi.com>

gherceg commented on corehq/project_limits/gauge.py:
```diff
@@ -0,0 +1,181 @@
+            # no throttling if config to throttle is not set.
+            return True
+        computed_value = None
+        if self.gauge_definition.actionable_metric == AVG:
```
```suggestion
        if self.gauge_definition.aggregator == AVG:
```

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/34800/commits/d980bd2fa35b0cfa4d635ae69bf8c2a4c3813392

AmitPhulera committed:
fix - missed variable renames in refactor

AmitPhulera committed:
add/update comments based on review

AmitPhulera committed:
add explicit None check for checking partition because partition numbers can be 0

AmitPhulera committed:
Merge branch 'master' into ap/gauge-limiter

AmitPhulera commented:
I have tested it on staging. I was able to get values of the `case_pillow_lag_gauge.max()` and `case_pillow_lag_gauge.avg()` which were same as the values observed on Datadog.

millerdev reviewed (commented):
LGTM. Not sure if all of Graham's concerns are addressed, so waiting for him to sign off.

AmitPhulera committed:
allow aggregrator to be None

since children classes can have their own logic of aggregations and can use multiple aggregrations at a time, so allow aggregator to be None
